### ISSUE TITLE
VO model training strategy 

### ISSUE THREAD

**Issue by AshwiniUthir** [Xiaoming-Zhao/PointNav-VO#11]
This is a follow up question to the issue "VO model training #10".
You mentioned that for VO training,
"For the 1-million-entry dataset we used in the paper, it needs 4-5 days to complete the whole training on a GeForce RTX 2080 Ti."

1. Does the 4-5 days training period include `move_forward`, `turn_left`, `turn_right`, and joint training for `turn_left`, `turn_right`  or for each action does it take 4-5 days? Can you please clarify?

2. Is there a way to train all the actions together ?

**Comment by Xiaoming-Zhao** [comment]
> Does the 4-5 days training period include move_forward, turn_left, turn_right, and joint training for turn_left, turn_right or for each action does it take 4-5 days?

Every action takes 4-5 days. If you have three GPUs, you can train them simultaneously.

> Is there a way to train all the actions together ?

For best performance, I recommend training them separately. See Row 6 of Tab.2 from [the paper](https://arxiv.org/pdf/2108.11550.pdf). Training all actions together yields inferior performance.

**Comment by AshwiniUthir** [comment]
Thank you for the quick response. For the VO model training, I have 140 checkpoints created so far. Can you please let me know how many checkpoints would get created with the default settings ?

**Comment by Xiaoming-Zhao** [comment]
By default, it will create 150 checkpoints:
https://github.com/Xiaoming-Zhao/PointNav-VO/blob/eb509b13234bba3ede81c1c5b96a46cf0c39e069/configs/vo/vo_pointnav.yaml#L40

I think 140 is enough. You can take a look at the validation curve in the tensorboard and choose the checkpoint with the lowest validation error.

**Comment by ruddradev** [comment]
Hello,

@Xiaoming-Zhao : Just wanted to reconfirm the checkpoint strategy, When I ran the code, I saw it creates 1 checkpoint every epoch, does that mean each action needs to be trained for 150 epochs on 1 million frames each? Or am I going wrong somewhere? Could you tell me the number of frames to train each action? Till now I can understand that if we train each action by X frames, it will take - 

X frame each for Left, Right and Forward actions
X frames again for Left, Right joint Geo invariance loss
14 million frames of finetuning for the RL checkpoint.

To reproduce the results of this paper, what would you suggest the value of X to be?

**Comment by ruddradev** [comment]
> Hello,
> 
> @Xiaoming-Zhao : Just wanted to reconfirm the checkpoint strategy, When I ran the code, I saw it creates 1 checkpoint every epoch, does that mean each action needs to be trained for 150 epochs on 1 million frames each? Or am I going wrong somewhere? Could you tell me the number of frames to train each action? Till now I can understand that if we train each action by X frames, it will take -
> 
> X frame each for Left, Right and Forward actions X frames again for Left, Right joint Geo invariance loss 14 million frames of finetuning for the RL checkpoint.
> 
> To reproduce the results of this paper, what would you suggest the value of X to be?

I just read the comment in another issue, that it actually is 150 checkpoints of 1 million each. According to the comment, it should take 1.5 hrs per epoch, however for me it is taking close to 22 hours on an A100, I need to find out what the reason could be. Is there anything that comes to your mind regarding this slowness?

**Comment by Xiaoming-Zhao** [comment]
@ruddradev
> To reproduce the results of this paper, what would you suggest the value of X to be?

Tab. S3 in the appendix of [the paper](https://arxiv.org/pdf/2108.11550.pdf) gives the X for each action. Actually, we have 1 million data in total for all three actions instead of 1 million per action.

The data comes from real trajectories, which can be generated as https://github.com/Xiaoming-Zhao/PointNav-VO/blob/main/TRAIN.md#dataset-generation

> it actually is 150 checkpoints of 1 million each

You do not need to run it for 150 epochs. You can monitor the learning curve in tensorboard and the training will converge before 150 epochs.

> I need to find out what the reason could be. Is there anything that comes to your mind regarding this slowness?

Is your GPU utilization high enough? If it is not, then it may be because of the slowness of the dataloader. You may need to timing each step in dataset. There may be some unsuitable configuration of the development environment. https://github.com/Xiaoming-Zhao/PointNav-VO/blob/eb509b13234bba3ede81c1c5b96a46cf0c39e069/pointnav_vo/vo/dataset/regression_geo_invariance_iter_dataset.py#L205
